rust/lsp-lib/src/lsp_plugin.rs

```diff
@@ -0,0 +1,458 @@
+use xi_plugin_lib::{Cache, ChunkCache, Plugin, View};
+use xi_rope::rope::RopeDelta;
+
+use types::Config;
+
+/// Represents the state of the Language Server Plugin
+pub struct LSPPlugin {
+    pub config: Config,
+    language_server_clients: HashMap<String, Arc<Mutex<LanguageServerClient>>>,
+}
+
+/// Get LSP Style Utf-16 based position given the xi-core style utf-8 offset
+fn get_position_of_offset<C: Cache>(
+    view: &mut View<C>,
+    offset: usize,
+) -> Result<Position, PluginLibError> {
+    let line_num = view.line_of_offset(offset)?;
+    let line_offset = view.offset_of_line(line_num)?;
+
+    let char_offset: usize = view.get_line(line_num)?[0..(offset - line_offset)]
+        .chars()
+        .map(char::len_utf16)
+        .sum();
+
+    Ok(Position {
+        line: line_num as u64,
+        character: char_offset as u64,
+    })
+}
+
+/// Get contents changes of a document modeled according to Language Server Protocol
+/// given the RopeDelta
+fn get_document_content_changes<C: Cache>(
+    delta: Option<&RopeDelta>,
+    view: &mut View<C>,
+) -> Result<Vec<TextDocumentContentChangeEvent>, PluginLibError> {
+    if let Some(delta) = delta {
+        if let Some(node) = delta.as_simple_insert() {
+            let (interval, _) = delta.summary();
+            let text = String::from(node);
+
+            let (start, end) = interval.start_end();
+            let text_document_content_change_event = TextDocumentContentChangeEvent {
+                range: Some(Range {
+                    start: get_position_of_offset(view, start)?,
+                    end: get_position_of_offset(view, end)?,
+                }),
+                range_length: Some((end - start) as u64),
+                text,
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+        // Or a simple delete
+        else if delta.is_simple_delete() {
+            let (interval, _) = delta.summary();
+
+            let (start, end) = interval.start_end();
+
+            // Hack around sending VSCode Style Positions to Language Server.
+            // See this issue to understand: https://github.com/Microsoft/vscode/issues/23173
+            let mut end_position = get_position_of_offset(view, end)?;
+
+            if end_position.character == 0 {
+                let mut ep = get_position_of_offset(view, end - 1)?;
+                ep.character += 1;
+                end_position = ep;
+            }
+
+            let text_document_content_change_event = TextDocumentContentChangeEvent {
+                range: Some(Range {
+                    start: get_position_of_offset(view, start)?,
+                    end: end_position,
+                }),
+                range_length: Some((end - start) as u64),
+                text: String::new(),
+            };
+
+            Ok(vec![text_document_content_change_event])
+        }
+        // Send the whole document again if it is not a trivial edit
+        else {
```

Comment:
I tried by early return but I am also using `?` operator, so that block can also return an Result::Err type so an else block is mandatory to return Result from that if-else expression.

Reply:
Hmm I'm confused. the `?` operator desugars to:

```rust
let _ = match result {
    Ok(result) => result,
    Err(err) return err.into(),
};
```

i.e, the `?` doesn't return an error from the block, it exits the function. So this should work, if it looks like:

```rust
 if let Some(delta) = delta {
        if let Some(node) = delta.as_simple_insert() {
            // ...
            return Ok(vec![text_document_content_change_event]);
        // Or a simple delete
        } else if delta.is_simple_delete() {
            // ...
            return Ok(vec![text_document_content_change_event]);
        }
    }
    // Send the whole document again if it is not a trivial edit
    let text_document_content_change_event = _;
    Ok(vec![text_document_content_change_event])
```

I think?

Reply:
Ahh! I realize. I did something like

```rust
 if let Some(delta) = delta {
        if let Some(node) = delta.as_simple_insert() {
            // ...
            return Ok(vec![text_document_content_change_event]);
        // Or a simple delete
        } else if delta.is_simple_delete() {
            // ...
            return Ok(vec![text_document_content_change_event]);
        }
    }
    // Send the whole document again if it is not a trivial edit
else { 
    let text_document_content_change_event = _;
    Ok(vec![text_document_content_change_event])
}
```

So the type inferred from `else` block was being expected from `if`. Fixed now 😄